### PR TITLE
Handle starting the frontend before finished loading integrations

### PR DESCRIPTION
**Pull request by bramkragten**
## Proposed change

For https://github.com/home-assistant/core/pull/36093

Needs https://github.com/home-assistant/home-assistant-js-websocket/pull/117

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/state/disconnect-toast-mixin.ts [hunk 1]
```diff
@@ -12,6 +12,31 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
...
+        !changedProperties.has("hass") ||
+        !oldHass?.config ||
+        oldHass.config.state === this.hass!.config.state
+      ) {
+        return;
+      }
+      if (this.hass!.config.state !== "RUNNING") {
+        showToast(this, {
+          message:
+            "Home Assistant is starting, not everything will be available untill it is finished.",
```
Comment (bdraco) [inline src/state/disconnect-toast-mixin.ts:28]:
```suggestion
            "Home Assistant is starting, not everything will be available until it is finished.",
```

File: src/layouts/partial-panel-resolver.ts [hunk 2]
```diff
@@ -128,6 +130,27 @@ class PartialPanelResolver extends HassRouterPage {
   private async _updateRoutes(oldPanels?: HomeAssistant["panels"]) {
     this.routerOptions = getRoutes(this.hass.panels);
 
+    if (
+      !this._waitForStart &&
+      this._currentPage &&
+      !this.hass.panels[this._currentPage]
+    ) {
+      if (this.hass.config.state !== "RUNNING") {
```
Comment (balloob) [inline src/layouts/partial-panel-resolver.ts:138]: `RUNNING` should be a constant.

File: src/layouts/partial-panel-resolver.ts [hunk 3]
```diff
@@ -128,6 +130,27 @@ class PartialPanelResolver extends HassRouterPage {
...
+        this._waitForStart = true;
+        window.addEventListener("connection-status", (ev) => {
```
Comment (balloob) [inline src/layouts/partial-panel-resolver.ts:140]: Seems weird we listen for it here, can't we read this from `hass` object?

File: src/panels/lovelace/common/generate-lovelace-config.ts [hunk 4]
```diff
@@ -468,6 +468,28 @@ export const generateLovelaceConfigFromHass = async (
   hass: HomeAssistant,
   localize?: LocalizeFunc
 ): Promise<LovelaceConfig> => {
+  if (hass.config.state !== "RUNNING") {
+    return {
+      title: hass.config.location_name,
+      views: [
+        {
+          cards: [{ type: "starting" }],
```
Comment (balloob) [inline src/panels/lovelace/common/generate-lovelace-config.ts:476]: Should this be a card or use `hass-loading-screen` instead ? That seems more in line with how we do loading screens.

File: src/state/connection-mixin.ts [hunk 5]
```diff
@@ -156,8 +157,25 @@ export const connectionMixin = <T extends Constructor<HassBaseEl>>(
 
     protected hassReconnected() {
       super.hassReconnected();
+
       this._updateHass({ connected: true });
       broadcastConnectionStatus("connected");
+
+      // @ts-ignore
+      this.hass!.callWS({ type: "get_config" }).then((config: HassConfig) => {
```
Comment (balloob) [inline src/state/connection-mixin.ts:165]: Don't we already call `get_config` ?

File: src/state/connection-mixin.ts [hunk 6]
```diff
@@ -156,8 +157,25 @@ export const connectionMixin = <T extends Constructor<HassBaseEl>>(
...
+        this._updateHass({ config });
+        if (config.state !== "RUNNING") {
+          const unsubProm = this.hass!.connection.subscribeEvents(() => {
```
Comment (balloob) [inline src/state/connection-mixin.ts:168]: Careful here. If Home Assistant finished starting before the subscription was in place, you will never know ! So you will need to fetch the config one more time once `unsubProm` resolves.

File: src/state/connection-mixin.ts (hunk 5 above)
Comment (balloob) [reply src/state/connection-mixin.ts:165]: oh nm this is `hassReconnected`.

Do we have this logic too for `hassConnected` ?

Comment (balloob) [reply src/state/connection-mixin.ts:165]: Actually, we use `subscribeConfig` for keeping the main config up to date, so we are already fetching the config there. Now we're fetching it twice ? 

I wonder if we should fire `core_config_updated` event in the backend when the state updates, so we can just listen to `config` updates via the subscription.

Comment (bramkragten) [reply src/state/connection-mixin.ts:165]: Yes, that was one of the options I was playing with, and that would also work, but the timing was difficult as you want the state right after the connect.

File: src/panels/lovelace/common/generate-lovelace-config.ts (hunk 4 above)
Comment (bramkragten) [reply src/panels/lovelace/common/generate-lovelace-config.ts:476]: Using `hass-loading-screen` will be a lot more work (we now just use Lovelace logic) and will not give a better experience I think?

File: src/state/connection-mixin.ts (hunk 5 above)
Comment (bramkragten) [reply src/state/connection-mixin.ts:165]: If the state would have updated before connect, you would never get it.

File: src/layouts/partial-panel-resolver.ts (hunk 2 above)
Comment (bramkragten) [reply src/layouts/partial-panel-resolver.ts:138]: Put those in `js-websocket`?

File: src/translations/en.json [hunk 7]
```diff
@@ -498,6 +498,8 @@
     "notification_toast": {
       "service_call_failed": "Failed to call service {service}.",
       "connection_lost": "Connection lost. Reconnecting…",
+      "started": "Home Assistant is started!",
```
Comment (bdraco) [inline src/translations/en.json:501]:
```suggestion
      "started": "Home Assistant has started!",
```

**Comment by bdraco** [comment]
Looking good, code is now catching the start

File: src/panels/lovelace/cards/hui-button-card.ts [hunk 8]
```diff
@@ -145,13 +145,9 @@ export class HuiButtonCard extends LitElement implements LovelaceCard {
...
-        <hui-warning
-          >${this.hass.localize(
-            "ui.panel.lovelace.warning.entity_not_found",
-            "entity",
-            this._config.entity
-          )}</hui-warning
-        >
+        <hui-warning>
+          ${createEntityNotFoundWarning(this.hass, this._config.entity)}
+        </hui-warning>
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-button-card.ts:150]: Why actually expect caller to still wrap it in `hui-warning` and not just return a template from `createEntityNotFoundWarning` ?

File: src/state/connection-mixin.ts (hunk 5 above)
Comment (balloob) [reply src/state/connection-mixin.ts:166]: The config is automatically refreshed when Home Assistant reconnects. That's built into the collections of Home Assistant Websocket JS: https://github.com/home-assistant/home-assistant-js-websocket/blob/master/lib/collection.ts#L57-L58

**Review by balloob — APPROVED** [review]

**Comment by balloob** [comment]
I wonder if we can split the backend PR to already add state, so we can unblock this PR?

**Comment by bdraco** [comment]
> I wonder if we can split the backend PR to already add state, so we can unblock this PR?

If I delete 
```
    startup_listeners.append(
        hass.bus.async_listen(EVENT_COMPONENT_LOADED, async_wait_frontend_load)
    )
```
from the backend PR, it should be mergeable asis.  Then we can add it back in later when supervisor is ready.

**Pull request merged**